woocommerce/woocommerce-gutenberg-products-block
# Disallow selecting quantity of stock past what's available.

This addresses #1887 

This pull does the following:

- `useStoreCartItemQuantity` now receives the cartItem instead of the `cartItemKey` as an argument. I didn't notice in previous reviews how it's used in the context where we already have a cartItem so implementing this reduces complexity and makes the hook more precise for it's purpose.
- Add `backorders_allowed` to the CartItem schema for the API. This allows for client to have correct logic for maximum quantity when this value is true.
- Implement the above in the `CartLineItemRow` so that quantity picker is limited by the amount of stock remaining if that is available (`lowStockRemaining`) and the `backorders_allowed` is false.
- maximum quantity is currently hardcoded to a (filtered) value of `99` when other conditions don't apply (see related issue in #1913)

## To Test

- Verify quantity selections work as expected.
- Verify if you have a product with stock set, and the amount remaining is within the lowStockRemaining threshold that you see the low stock badge and the quantity is limited by the amount remaining.
- Verify that if the above is true but the product also has `backorders_allowed` set to true that the quantity limit is 99.
- Verify for all other conditions the quantity limit maxes out at 99.

## Potential followup.

Should we automatically reduce the quantity selected for an item in the cart if the stock has gone down? Since we're updating the cart on every request, it's possible for us to do this (and we could trigger a notice when this happens). This could entice folks to move quicker on high turnover items.

We could do the above as a stretch goal, but it's nice knowing it's doable.
